x/incentives/keeper/distribution.go

```diff
@@ -4,87 +4,108 @@ import (
 	"math/big"
 	"strconv"
 
+	"github.com/armon/go-metrics"
+	"github.com/cosmos/cosmos-sdk/telemetry"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 	"github.com/ethereum/go-ethereum/common"
 
 	"github.com/tharsis/evmos/v4/x/incentives/types"
 )
 
-// Distribute transfers the allocated rewards to the participants of a given
+// DistributeRewards transfers the allocated rewards to the participants of a given
 // incentive.
 //  - allocates the amount to be distributed from the inflation pool
 //  - distributes the rewards to all particpants
 //  - deletes all gas meters
 //  - updates the remaining epochs of each incentive
 //  - sets the cumulative totalGas to zero
-func (k Keeper) DistributeIncentives(ctx sdk.Context) error {
+func (k Keeper) DistributeRewards(ctx sdk.Context) error {
 	logger := k.Logger(ctx)
 
-	// Allocate rewards for each Incentive
-	coinsAllocated, err := k.allocateCoins(ctx)
+	rewardAllocations, rewardsTotal, err := k.rewardAllocations(ctx)
 	if err != nil {
 		return err
 	}
 
-	// Iterate over each incentive and distribute allocated rewards
-	k.IterateIncentives(
-		ctx,
-		func(incentive types.Incentive) (stop bool) {
-			// Distribute rewards
-			k.rewardParticipants(ctx, incentive, coinsAllocated)
-
-			// Update epoch
-			incentive.Epochs--
-
-			// Update Incentive and reset its total gas count
-			if incentive.IsActive() {
-				k.SetIncentive(ctx, incentive)
-				k.SetIncentiveTotalGas(ctx, incentive, 0)
-			} else {
-				// Remove incentive if it has no remaining epochs
-				k.DeleteIncentiveAndUpdateAllocationMeters(ctx, incentive)
-				logger.Info(
-					"incentive finalized",
-					"contract", incentive.Contract,
+	var participants uint64
+	k.IterateIncentives(ctx, func(incentive types.Incentive) (stop bool) {
+		rewards, count := k.rewardParticipants(ctx, incentive, rewardAllocations)
+		participants += count
```

Comment:
it's worth noticing that these might not all be unique addresses

Reply:
@fedekunze they should be unique addresses though cause participants are counted by gasMeters in `rewardParticipants`